probe_scraper/glean_checks.py

```diff
@@ -0,0 +1,81 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+This file contains various sanity checks for Glean.
+"""
+
+import os
+
+from schema import And, Optional, Schema
+
+
+def check_glean_metric_structure(data):
+    schema = Schema(
+        {
+            str: {
+                Optional(And(str, lambda x: len(x) == 40)): [
+                    And(str, lambda x: os.path.exists(x))
+                ]
+            }
+        }
+    )
+
+    schema.validate(data)
+
+
+def check_for_duplicate_metrics(repositories, metrics_by_repo, emails):
+    """
+    Checks for duplicate metric names across all libraries used by a particular application.
+    Queues a warning e-mail if any are found, and removes all metrics for the app with
+    duplicate metrics.
+    """
+    repo_by_library_name = {}
+    repo_by_name = {}
+    for repo in repositories:
+        for library_name in repo.library_names or []:
+            repo_by_library_name[library_name] = repo.name
+        repo_by_name[repo.name] = repo
+
+    for repo in repositories:
+        dependencies = [repo.name] + [
+            repo_by_library_name[library_name] for library_name in repo.dependencies
+        ]
+
+        metric_sources = {}
+        for dependency in dependencies:
+            for metric in metrics_by_repo[dependency].keys():
+                metric_sources.setdefault(metric, []).append(dependency)
+
+        duplicate_sources = dict(
+            (k, v) for (k, v) in metric_sources.items() if len(v) > 1
+        )
+
+        if len(duplicate_sources):
+            for name, sources in duplicate_sources.items():
```

Comment:
We _could_ send a single email with all dupes, but that wouldn't really apply to probe-owners (who aren't repo owners).

Reply:
I'd really expect dupes to be a low frequency exception, not the rule: maybe receiving a few emails about it isn't that bad

Reply:
I feel like the best compromise here might be to: Send a single e-mail with all dupes, that goes to all repo owners and all metric owners involved.  It does mean that the e-mail will contain metrics that some of the metric owners don't care about, but given that there will be other metrics in there that they *do* care about, it seems like a small price to pay.  And I think I like that better than the current behavior which has the potential (in rare cases) to send many e-mails to the same individual in a single run.